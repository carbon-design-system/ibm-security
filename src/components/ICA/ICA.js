/**
 * @file Important content area (ICA).
 * @copyright IBM Security 2019 - 2020
 */

import classnames from 'classnames';
import numeral from 'numeral';
import PropTypes from 'prop-types';
import React from 'react';

import 'numeral/locales';

import { ArrowUp16, ArrowUp20, ArrowUp24 } from '@carbon/icons-react';
import Icon from '../Icon/Icon';

import isDevelopment from '../../globals/env';
import { getComponentNamespace } from '../../globals/namespace';

export const namespace = getComponentNamespace('ica');

/**
 * @type {string[]} Array of accepted locales by Numeral.js.
 */
export const Locales = Object.keys(numeral.locales);

/**
 * Ensure that the format string includes a decimal space only for values that
 * are 1000 and above.
 * @param {number} value The value to evaluate.
 * @returns {string} Format string for numeral.
 */
const getFormat = value => (Math.round(value) > 999 ? '0.0a' : '0a');

/**
 * Ensure that the value is formatted correctly based on whether it should be truncated or not.
 * @param {number} value The value to format.
 * @param {boolean} truncate Whether or not the value should be truncated.
 * @returns {string} The formatted value.
 */
const formatValue = (value, truncate) => {
  const localeValue = numeral(value);

  return truncate ? localeValue.format(getFormat(value)) : localeValue.value();
};

/**
 * Ensure that the value includes a percentage if specified by prop,
 * or otherwise the value is properly formatted.
 * @param {boolean} percentage Whether a percent sign should be included.
 * @param {number|null} value The value to be formatted.
 * @param {boolean} truncate Whether or not the value should be truncated.
 * @returns {string} Formatted string.
 */
const truncateValue = (percentage, value, truncate) => {
  if (percentage)
    return (
      <div className={`${namespace}__percentage`}>
        {value}
        <span className={`${namespace}__percentage-mark`}>%</span>
      </div>
    );

  return value === null ? 'â€“' : formatValue(value, truncate);
};

const ICA = ({
  className,
  forceShowTotal,
  label,
  locale,
  percentage,
  size,
  trending,
  truncate,
  total,
  value,
  ...other
}) => {
  const isSize = sizeValue => size === sizeValue;
  const isLarge = isSize('lg');
  const isXLarge = isSize('xl');

  let renderIcon = ArrowUp16;
  if (isLarge) {
    renderIcon = ArrowUp20;
  } else if (isXLarge) {
    renderIcon = ArrowUp24;
  }

  const ICAClasses = classnames(className, {
    [`${namespace}--lg`]: isLarge,
    [`${namespace}--xl`]: isXLarge,
  });

  if (Locales.includes(locale)) {
    numeral.locale(locale);
  } else {
    if (isDevelopment()) {
      console.warn(
        `Locale "${locale}" for ICA component is not recognized. Using fallback locale "${ICA.defaultProps.locale}. For list of supported locales go to: https://github.com/carbon-design-system/ibm-security/tree/master/src/components/ICA`
      );
    }
    numeral.locale(ICA.defaultProps.locale);
  }

  const truncatedValue = truncateValue(percentage, value, truncate);
  const truncatedTotal = formatValue(total, truncate);

  const shouldDisplayof =
    (!percentage && total > value && truncatedValue !== truncatedTotal) ||
    (forceShowTotal && truncatedTotal > 0);

  return (
    <div className={`${ICAClasses}`} {...other}>
      <h4 className={`${namespace}__label`}>{label} </h4>
      <span className={`${namespace}__row`}>
        {trending && (
          <Icon className={`${namespace}__trend`} renderIcon={renderIcon} />
        )}
        <span className={`${namespace}__value`}>{truncatedValue}</span>
        {shouldDisplayof ? (
          <span className={`${namespace}__total`}>
            {' '}
            <span>/{truncatedTotal}</span>
          </span>
        ) : null}
      </span>
    </div>
  );
};

ICA.propTypes = {
  /**
   * Text label for ICA.
   * @type string
   */
  label: PropTypes.string.isRequired,

  /**
   * The main ICA value to display
   * @type number
   */
  value: PropTypes.number,

  /**
   * Total value that the main ICA value is a subset of.
   * @type number
   */
  total: PropTypes.number,

  /**
   * Optional class name.
   * @type number
   */
  className: PropTypes.string,

  /**
   * Locale value to determine approach to formatting numbers.
   * @type string
   */
  locale: PropTypes.oneOf(Locales),

  /**
   * Format number to percentage when `percentage` prop is true.
   * @type bool
   */
  percentage: PropTypes.bool,

  /**
   * Display the `total` even when the `value` is equal to
   * the `total` when `forceShowTotal` prop is true on the
   * condition that the `total` is greater than 0.
   * @type bool
   */
  forceShowTotal: PropTypes.bool,

  /** Specify whether or not the values should be truncated. */
  truncate: PropTypes.bool,

  /** The size of the ICA. */
  size: PropTypes.oneOf(['default', 'lg', 'xl']),

  /** Display trending icon. */
  trending: PropTypes.bool,
};

ICA.defaultProps = {
  total: 0,
  className: '',
  locale: 'en',
  percentage: false,
  size: 'default',
  value: null,
  forceShowTotal: false,
  trending: false,
  truncate: true,
};

export default ICA;
