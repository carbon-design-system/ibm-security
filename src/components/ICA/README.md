# Important Content Area (ICA) component

## Value truncation

ICA components provide a simple way to display numerical values. Even for long numbers, this component truncates these numbers so they render as short as possible. It follows different language conventions to do so and consuming developers are expected to provide the locale code for their users for the component to truncate correctly. The default locale is English (`en`) if one is not provided or if the provided one is not supported.

### Supported locales for value truncation

| Code    | Locale                 |
| ------- | ---------------------- |
| `bg`    | Bulgarian              |
| `chs`   | Simplified Chinese     |
| `cs`    | Czech                  |
| `da-dk` | Danish Denmark         |
| `de-ch` | German in Switzerland  |
| `de`    | German                 |
| `en-au` | English Australia      |
| `en-gb` | English United Kingdom |
| `en-za` | English South Africa   |
| `es-es` | Spanish Spain          |
| `es`    | Spanish                |
| `et`    | Estonian               |
| `fi`    | Finnish                |
| `fr-ca` | French Canada          |
| `fr-ch` | French Switzerland     |
| `fr`    | French                 |
| `hu`    | Hungarian              |
| `it`    | Italian                |
| `ja`    | Japanese               |
| `lv`    | Latvian                |
| `nl-be` | Dutch Belgium          |
| `nl-nl` | Dutch Netherlands      |
| `no`    | Norwegian              |
| `pl`    | Polish                 |
| `pt-br` | Portuguese Brazil      |
| `pt-pt` | Portuguese Portugal    |
| `ru-ua` | Russian Ukraine        |
| `ru`    | Russian                |
| `sk`    | Slovak                 |
| `sl`    | Slovenian              |
| `th`    | Thai                   |
| `tr`    | Turkish                |
| `uk-ua` | Ukrainian              |
| `vi`    | Vietnamese             |
