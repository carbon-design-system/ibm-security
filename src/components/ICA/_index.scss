////
/// Important content area component.
/// @group ica
/// @copyright IBM Security 2019 - 2021
////

@import '@carbon/type/scss/styles';
@import '@carbon/themes/scss/tokens';

@import 'ICASkeleton/index';

@import '../Component/mixins';

@include security--component($name: ica) {
  &__label {
    @include carbon--type-style($name: label-01);

    display: block;
    margin: 0;
    color: $text-02;
  }

  &__value {
    @include carbon--type-style($name: productive-heading-04);

    margin-bottom: 0;
  }

  &__total {
    @include carbon--type-style($name: body-short-01);

    padding-top: $carbon--spacing-01 + $carbon--spacing-04;
    margin-bottom: 0;
  }

  &__trend {
    margin-top: $carbon--spacing-03;
    vertical-align: top;
  }

  &__row {
    display: flex;
  }

  &__percentage {
    @include carbon--type-style($name: productive-heading-04);
  }

  &__percentage-mark {
    @include carbon--type-style($name: body-short-01);
  }

  // Lg Styling
  &--lg & {
    &__value,
    &__percentage {
      @include carbon--type-style($name: productive-heading-06);
    }

    &__total,
    &__percentage-mark {
      @include carbon--type-style($name: productive-heading-03);

      padding-top: $carbon--spacing-03 + $carbon--spacing-04;
    }
  }

  //  Styling
  &--xl & {
    &__label {
      @include carbon--type-style($name: body-short-01);
    }

    &__value,
    &__percentage {
      @include carbon--type-style($name: productive-heading-07);
    }

    &__total,
    &__percentage-mark {
      @include carbon--type-style($name: productive-heading-04);

      padding-top: $carbon--spacing-06;
    }

    &__trend {
      margin-top: $carbon--spacing-04;
    }
  }
}
