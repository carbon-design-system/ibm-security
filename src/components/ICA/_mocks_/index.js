/**
 * @file Important content area mocks.
 * @copyright IBM Security 2019
 */

const value = 10;

export default {
  label: 'Label',
  value,
  total: value * value,
};
