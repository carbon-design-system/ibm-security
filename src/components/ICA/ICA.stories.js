/**
 * @file Important content area stories.
 * @copyright IBM Security 2019 - 2021
 */

import { boolean, select, number, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import React from 'react';

import { components } from '../../../.storybook';
import { carbonPrefix } from '../../globals/namespace';

import { ICA } from '../..';
import { Locales } from './ICA';

import props from './_mocks_';

const { label, total, value: icaValue } = props;

const sizes = {
  Default: 'default',
  Large: 'lg',
  XLarge: 'xl',
};

const storyProps = ({ value = icaValue, total } = {}) => ({
  label: text('Label (label)', label),
  value: number('Value (value)', value),
  total: number('Total (total)', total),
  percentage: boolean('Percentage (percentage)', false),
  size: select('ICA size (size)', sizes, 'default'),
  forceShowTotal: boolean('Show total (forceShowTotal)', false),
  trending: boolean('Show trending arrow (trending)', false),
  truncate: boolean('Truncate (truncate)', ICA.defaultProps.truncate),
  locale: select('Locale (locale)', Locales, Locales[0]),
});

storiesOf(components('ICA'), module)
  .add('default', () => <ICA {...storyProps()} />)
  .add('with null value', () => <ICA {...storyProps({ value: null })} />)
  .add('with 1000 value', () => (
    <ICA {...storyProps({ value: icaValue * 100 })} />
  ))
  .add('with 1000000 value', () => (
    <ICA {...storyProps({ value: icaValue * 100000 })} />
  ))
  .add('with total', () => <ICA {...storyProps({ total })} />)
  .add(
    'in an ICA wall',
    () => (
      <div
        className={`${carbonPrefix}--grid ${carbonPrefix}--grid--full-width`}
      >
        <div className={`${carbonPrefix}--row`}>
          <div className={`${carbonPrefix}--col`}>
            <h4>4 spaced</h4>
          </div>
        </div>
        <div className={`${carbonPrefix}--row`}>
          {Array(4)
            .fill(0)
            .map(item => (
              <div
                key={item.id}
                className={`${carbonPrefix}--col-sm-4 ${carbonPrefix}--col-md-2 ${carbonPrefix}--col-lg-4`}
              >
                <ICA {...storyProps({ total })} />
              </div>
            ))}
        </div>
        <div className={`${carbonPrefix}--row`}>
          <div className={`${carbonPrefix}--col`}>
            <h4>8 condensed</h4>
          </div>
        </div>
        <div className={`${carbonPrefix}--row`}>
          {Array(8)
            .fill(0)
            .map(item => (
              <div
                key={item.id}
                className={`${carbonPrefix}--col-sm-2 ${carbonPrefix}--col-md-2 ${carbonPrefix}--col-lg-2`}
              >
                <ICA {...storyProps({ total })} />
              </div>
            ))}
        </div>
      </div>
    ),
    {
      info: {
        text: `
        Multiple \`ICA\` components (i.e., an "ICA Wall") should be presented in a grid using the correct class names.

        These two row examples show different combinations of breakpoints and spans set per column with specific class names.
        
        For more information the 16 column IBM grid, please review the [\`@carbon/grid\` package documentation](https://github.com/carbon-design-system/carbon/tree/main/packages/grid).
      `,
      },
    }
  );
