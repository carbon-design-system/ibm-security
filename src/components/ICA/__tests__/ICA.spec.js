/**
 * @file Important content area tests.
 * @copyright IBM Security 2019
 */

import React from 'react';
import { render } from '@testing-library/react';

import { ICA } from '../../..';
import { Locales } from '../ICA';

describe('ICA', () => {
  test('should have no Axe or DAP violations', async () => {
    const main = document.createElement('main');
    render(<ICA label="test ICA" total={10} value={5} />, {
      // DAP requires a landmark '<main>' in the DOM:
      container: document.body.appendChild(main),
    });
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations('ICA');
  });

  it('should render en dash when `value` is `null`', () => {
    const { queryByText } = render(<ICA label="test ICA" value={null} />);
    expect(queryByText('â€“')).toBeVisible();
  });

  it('should render a large value', () => {
    const { queryByText } = render(<ICA label="test ICA" value={1000000} />);
    expect(queryByText('1.0m')).toBeVisible();
  });

  it('should render a subset of values when `value` and `total` are provided', () => {
    const { getByText } = render(<ICA label="test ICA" value={5} total={10} />);
    expect(getByText('5').nextSibling).toBe(getByText(/10/i));
  });

  it('should not render subset of values when `total` is close to `value`', () => {
    const { queryByText } = render(
      <ICA label="test ICA" value={1000000} total={999995} />
    );
    expect(queryByText('999995')).not.toBeInTheDocument();
    expect(queryByText('1.0m')).toBeVisible();
  });

  it('should not render `total` if it is the same as `value`', () => {
    const { getByText } = render(
      <ICA label="test ICA" value={10} total={10} />
    );
    // Expect 2 ICA nodes: the label and the value.
    expect(getByText(/test ICA/i).parentNode.children.length).toBe(2);
  });

  it('should render `total` when forced', () => {
    const { getByText } = render(
      <ICA label="test ICA" value={10} total={10} forceShowTotal />
    );
    // Expect 3 ICA nodes: the label, value, and the total.
    expect(getByText(/test ICA/i).parentNode.children.length).toBe(3);
  });

  it('should render a percentage', () => {
    const { queryByText } = render(
      <ICA label="test ICA" value={10} percentage />
    );
    expect(queryByText('10').firstElementChild).toBe(queryByText('%'));
  });

  it('should render a percentage', () => {
    const { queryByText } = render(
      <ICA label="test ICA" value={10} percentage />
    );
    expect(queryByText('10').firstElementChild).toBe(queryByText('%'));
  });

  Locales.forEach(locale =>
    test(`should accept '${locale}' locale`, () => {
      const { container } = render(
        <ICA label="test ICA" value={10} locale={locale} />
      );
      expect(() => container).not.toThrow();
    })
  );
});
