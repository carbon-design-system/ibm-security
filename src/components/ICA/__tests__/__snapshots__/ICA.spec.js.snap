// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ICA does not render a subset when the total is close to the provided value 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    1.0k
  </span>
</div>
`;

exports[`ICA renders 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    10
  </span>
</div>
`;

exports[`ICA renders a large value 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    1.0k
  </span>
</div>
`;

exports[`ICA renders a percentage 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    <div
      className="security--ica__percentage"
    >
      10
      <sup
        className="security--ica__percentage-mark"
      >
        %
      </sup>
    </div>
  </span>
</div>
`;

exports[`ICA renders a subset of values 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    10
  </span>
  <span
    className="security--ica__total"
  >
     / 
    100
  </span>
</div>
`;

exports[`ICA renders a total when forced 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    10
  </span>
  <span
    className="security--ica__total"
  >
     / 
    10
  </span>
</div>
`;

exports[`ICA renders the null value 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    â€“
  </span>
</div>
`;

exports[`ICA renders with a value that is same as the total value 1`] = `
<div
  className="security--ica "
>
  <h4
    className="security--ica__label"
  >
    Label
     
  </h4>
  <span
    className="security--ica__value"
  >
    10
  </span>
</div>
`;
